Add node groups to opcua input plugin

reimda opened the pull request:
Remove unneeded node settings (data_type and description).

Add the concept of a node group.  Groups have default settings for all nodes in the group.  They also are able to override the plugin's metric name so multiple metric names can be used in a single plugin instance.

Remove the name tag which wasn't useful because it duplicated the OPC UA value's field name.

Expose stats to telegraf internal plugin.

ssoroka commented on plugins/inputs/opcua/README.md:
```diff
@@ -47,34 +47,57 @@ Plugin minimum tested version: 1.16
-  ## identifier			- tag as shown in opcua browser
-  ## data_type  			- boolean, byte, short, int, uint, uint16, int16,
-  ##                        uint32, int32, float, double, string, datetime, number
+  ## field_name        - field name to use in the output
```
this documentation is a bit confusing; I thought these were top-level settings until I got to the example below. Not sure what we can do to improve it.

reimda replied:
Nodes was a top level setting before this PR.  I added groups in this PR and each group also has a set of nodes.  I left in the top level nodes for backward compatibility.  I can see that this is confusing so I'll have to work on the docs.

ssoroka commented on plugins/inputs/opcua/opcua_client.go:
```diff
@@ -48,13 +49,24 @@ type OpcUA struct {
-type OPCTag struct {
-	Name           string `toml:"name"`
+type NodeSettings struct {
+	FieldName      string `toml:"field_name"`
```
backwards support for name might be good here.

reimda replied:
done

ssoroka commented on plugins/inputs/opcua/opcua_client.go:
```diff
@@ -13,11 +13,12 @@ import (
 // OpcUA type
 type OpcUA struct {
-	Name           string          `toml:"name"`
+	MetricName     string          `toml:"metric_name"`
```
backwards support for name might be good here.

reimda replied:
I wasn't worried about backward compatibility when I started this PR because I thought this plugin was new enough and broken enough that there wouldn't be enough existing users to worry about.  Your and Sven's concerns made me change my mind so I've reverted this.

ssoroka commented on plugins/inputs/opcua/opcua_client.go:
```diff
@@ -48,13 +49,24 @@ type OpcUA struct {
 	Namespace      string `toml:"namespace"`
 	IdentifierType string `toml:"identifier_type"`
 	Identifier     string `toml:"identifier"`
-	DataType       string `toml:"data_type"`
-	Description    string `toml:"description"`
```
removing these will make old configs fail. A concern?

reimda replied:
I added them back for compatibility but removed them from the readme because they were unnecessary and were never actually used.

ssoroka reviewed (commented):
👍🏼

srebhan commented on plugins/inputs/opcua/opcua_client.go:
```diff
@@ -392,16 +441,15 @@ func (o *OpcUA) Gather(acc telegraf.Accumulator) error {
-		fields["Quality"] = strings.TrimSpace(fmt.Sprint(o.NodeData[i].Quality))
-		acc.AddFields(o.Name, fields, tags)
+		fields[o.nodeData[i].TagName] = o.nodeData[i].Value
+		fields["quality"] = strings.TrimSpace(fmt.Sprint(o.nodeData[i].Quality))
```
For case-sensitive downstream applications (such as InfluxDB IIRC) this might be a problem. First it will generate new series AFAIK as field names are used for hashing **and** it might break queries further down...
Should we really risk this breakage?

reimda replied:
Telegraf fields are usually lower case.  I wasn't worried about backward compatibility when I changed this, but you have valid concerns.  reverted.

reimda commented:
I reverted the changes that affect backward compatibility.  I think you both are right that this plugin has been released long enough that we should keep compatibility.

I also added a feature OPC UA users have requested, which is to be able to add tags to specific node IDs or to groups.  Sorry for sneaking this feature into this PR after the initial review but it seemed closely related and not big enough to start a separate PR.

srebhan commented on plugins/inputs/opcua/opcua_client.go:
```diff
@@ -210,50 +312,75 @@ func (o *OpcUA) InitNodes() error {
-	for i, item := range o.NodeList {
+	nameEncountered := map[metricParts]struct{}{}
+	for i := range o.nodes {
+		node := &o.nodes[i]
```
Couldn't you get this directly in the for loop above?

reimda replied:
fixed, thanks

srebhan commented on plugins/inputs/opcua/opcua_client.go:
```diff
@@ -210,50 +312,75 @@ func (o *OpcUA) InitNodes() error {
+	tags       string // sorted by tag name and in format tag1=value1, tag2=value2
+}
+
+func newMP(n *Node) metricParts {
+	var keys []string
+	for key := range n.metricTags {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+	var sb strings.Builder
+	for i, key := range keys {
+		if i != 0 {
+			sb.WriteString(", ")
+		}
+		sb.WriteString(key)
+		sb.WriteString("=")
+		sb.WriteString(n.metricTags[key])
+	}
+	x := metricParts{
+		metricName: n.metricName,
+		fieldName:  n.tag.FieldName,
+		tags:       sb.String(),
+	}
+	return x
+}
```
This function might benefit speedwise from the optimization done here: https://github.com/influxdata/telegraf/pull/8391.

reimda replied:
Since this validation is only needed once during plugin initialization I didn't think optimization was worthwhile

srebhan reviewed (approved):
Two very minor comments. Looks good to me.
